src/bartiq/compilation/_compile.py

```diff
@@ -11,715 +11,234 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
-import warnings
-from typing import Any, Optional, cast, overload
-
-from .. import Port, Routine
+from __future__ import annotations
+
+from collections import defaultdict
+from collections.abc import Iterable
+from dataclasses import dataclass, replace
+from graphlib import TopologicalSorter
+from typing import Generic, TypeVar
+
+from qref import SchemaV1
+from qref.schema_v1 import RoutineV1
+from qref.verification import verify_topology
+
+from .._routine import (
+    CompiledRoutine,
+    Constraint,
+    ConstraintStatus,
+    Endpoint,
+    Port,
+    Routine,
+    routine_to_qref,
+)
 from ..errors import BartiqCompilationError
-from ..precompilation._core import PrecompilationStage, precompile
-from ..routing import get_port_source, get_port_target, join_paths
 from ..symbolics import sympy_backend
-from ..symbolics.backend import SymbolicBackend, T_expr
-from ..symbolics.variables import DependentVariable
-from ..verification import verify_compiled_routine, verify_uncompiled_routine
-from ._symbolic_function import (
-    RoutineWithFunction,
-    SymbolicFunction,
-    compile_functions,
-    define_expression_functions,
-    rename_functions,
-    rename_variables,
-    to_symbolic_function,
-    update_routine_with_symbolic_function,
-)
-from ._utilities import get_children_in_walk_order, is_constant_int, is_single_parameter
-from .types import FunctionsMap, Number
+from ..symbolics.backend import ComparisonResult, SymbolicBackend, TExpr
+from ._common import evaluate_ports, evaluate_resources
+from .preprocessing import DEFAULT_PRECOMPILATION_STAGES, PreprocessingStage
 
+T = TypeVar("T")
 
-@overload
-def compile_routine(
-    routine: Routine,
-    *,
-    precompilation_stages: Optional[list[PrecompilationStage]] = None,
-    global_functions: Optional[list[str]] = None,
-    functions_map: Optional[FunctionsMap] = None,
-    skip_verification: bool = False,
-) -> Routine:
-    pass  # pragma: no cover
-
+ParameterTree = dict[str | None, dict[str, TExpr[T]]]
```

Comment:
Would be nice to have a comment explaining what it represents?

Reply:
I added a comment explaining this data structure in [075ea09](https://github.com/PsiQ/bartiq/pull/115/commits/075ea09b02a6bae5846a26905655e6c39729078d). WDYT?

Reply:
Much better, thanks!